### ISSUE TITLE
Rule proposal: Forbid `...(foo || {})` and `...(foo ?? {})` in object literal

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#1475]
Unlike spread in array literals, spread falsy values in object won't throw error, and works as expected. There is no need to add empty object `{}` as  fallback. 

## Fail

```js
const object = {
	foo,
	...(bar || {}),
};
```

```js
const object = {
	foo,
	...(bar ?? {}),
};
```

## Pass

```js
const object = {
	foo,
	...bar,
};
```

```js
const array = [
	foo,
	...(bar || []),
];
```

Not sure about the rule name, `prefer-directly-spread`? `no-useless-empty-object`?

<details>
<summary>Spread falsy value</summary>

```js
[
	  0,
	  false,
	  undefined,
      '',
	  NaN,
	  0n,
	  null,
	  document.all, // The only exception, but who care?
].map(value => Object.keys(({...value})).length)
```

</details>

**Comment by sindresorhus** [comment]
`no-useless-fallback-in-spread`?

**Comment by fisker** [comment]
Sounds good

**Comment by fisker** [comment]
This is now accepted.

**Comment by will-stone** [comment]
Hi @fisker Thanks for another great rule. I've found that TS doesn't like spreading of falsy values, producing a "Spread types may only be created from object types" error.

<img width="761" alt="image" src="https://user-images.githubusercontent.com/654103/134761629-b6b58549-970d-4b85-8b3b-0437f9f8917a.png">

Not sure if that affects this package or not; if you're not worried about TS. I can simply turn the rule off. However I thought I'd point it out just in case.

https://github.com/microsoft/TypeScript/issues/37491

**Comment by fisker** [comment]
Thanks for the information, but I don't
 use ts myself, I'd like hear other ts users opinion.

**Comment by fisker** [comment]
Thanks for the information, but I don't
 use ts myself, I'd like hear other ts users opinion.

**Comment by papb** [comment]
@will-stone Your `a` in the screenshot has type `undefined`, which means it will always be `undefined`. If you create a variable with type `{ foo: "bar" } | undefined`, then TS will not complain:

```ts
const x = Math.random() < 0.5 ? { foo: 123 } : undefined;
const y = { ...x }; // Works fine
```

[See here](https://github.com/microsoft/TypeScript/issues/37491#issuecomment-601741064)

**Comment by will-stone** [comment]
Thank you @papb, that makes sense 🙂 However, I'm sure I noticed it from a real world example of my issue. I'll turn the rule back on in my shared eslint config and see if I can find the project it affected. If not, then happy days!
